sim-lib/src/lib.rs

```diff
@@ -553,6 +627,84 @@ async fn produce_events(
     shutdown.trigger();
 }
 
+async fn produce_random_events<N: NetworkGenerator, A: PaymentGenerator + Display>(
+    source: PublicKey,
+    network_generator: Arc<Mutex<N>>,
+    mut node_generator: A,
+    sender: Sender<SimulationEvent>,
+    shutdown: Trigger,
+    listener: Listener,
+) {
+    log::info!("Started random activity producer for {source}: {node_generator}.");
+
+    loop {
+        let wait = node_generator.next_payment_seconds();
+        log::trace!("Next payment for {source} in {:?} seconds.", wait);
+
+        select! {
+            biased;
+            _ = listener.clone() => {
+                break;
+            },
+
+            // Wait until our time to next payment has elapsed then execute a random amount payment to a random
+            // destination.
+            _ = time::sleep(wait) => {
+                let mut generator = network_generator.lock().await;
+
+                let destination = {
+                    let mut payment_destination = None;
+                    for _ in 0..3 {
+                        let dest = generator.sample_node_by_capacity();
+                        if dest.0 != source {
+                            payment_destination = Some(dest);
+                            break;
+                        }
+                        log::trace!("Re-selecting destination that is not self: {source}.");
+                    }
+
+                    payment_destination
+                };
+
+                // Only proceed with a payment if we managed to pick a destination that is not ourselves, otherwise
+                // skip this round.
+                if destination.is_none(){
+                    log::debug!("Skipping payment: could not select a destination that is not our own node: {source}.");
+                    continue;
+                }
+                let destination = destination.unwrap();
+
+                // Only proceed with a payment if the amount is non-zero, otherwise skip this round.
+                let amount = node_generator.payment_amount(destination.1);
+                if let Err(e) = amount{
+                    log::error!("Could not get amount for {source} -> {} activity: {e}.", destination.0);
```

Comment:
>  think we should note this here so we are able to tweak the random generator values if this happens too often.

Added more specific logging so that this is easier to debug. 

> Also, notice that the odds of this happening depend on whether the expected payment value is geq than the payment limit over the selected source and destination.

Do you think it's worth excluding nodes that have capacity < expected payment amount (or logging a warning)? Can see that being a real frustration running something like this, so a log noting that you need to update that value would be very helpful IMO.

Reply:
Exclude should be fine given if a node has `capacity < 2*expected_payment_amt` it will always fail to be paired with any other node